lnst/Recipes/ENRT/helpers.py

```diff
@@ -29,3 +30,17 @@ def ip_endpoint_pairs(
                 )
 
     return endpoint_pairs
+
+
+TIPEndpointPair = TypeVar("TIPEndpointPair", bound=EndpointPair[IPEndpoint])
+
+def filter_ip_endpoint_pairs(ip_versions: Collection[str], endpoint_pairs: Collection[TIPEndpointPair]) -> Iterator[TIPEndpointPair]:
+    def ip_version_string(ip_address: BaseIpAddress) -> str:
+        return "ipv4" if isinstance(ip_address, Ip4Address) else "ipv6"
+
+    for ip_version in ip_versions:
```

Comment:
This was intentional. I felt that first filtering by IP type, then doing product, was much cleaner than doing a product over all IPs of both endpoints end returning only endpoint pairs where both IPs are of the same type.

We shouldn't rely on the returned endpoints to be sorted in any way. Should I switch to first doing product, then filtering? Is there a good reason for it?

Reply:
this function doesn't do the product of endpoints though... it already accepts the list of endpoints and then returns a filtered list of endpoints. so it serves as a generic filter function for a list of objects. and imo that sort of filtering should be STABLE wrt. how the input list was ordered.

Reply:
Ah, I thought we're talking about the `ping_endpoint_pairs` function and not `filter_ip_endpoint_pairs`. Sure, I agree the filter should be stable.

The reason for this is to display ipv4 and ipv6 endpoints one after the other in results/logs since all are run in parallel, but this ordering is already done in `ping_endpoint_pairs`. I'll change the filter to be stable.